doc/Memory_Layout.md

```diff
@@ -1,75 +1,112 @@
-Memory Layout
+Memory Layout and Isolation
 =============
 
-This document describes how the memory in Tock is used for the
-kernel, applications, and supporting state.
+This document describes how the memory in Tock is structured and used for the
+kernel, applications, and supporting state.  It also describes how memory is
+isolated, meaning the access permissions of the kernel and processes.
 
 <!-- npm i -g markdown-toc; markdown-toc -i Memory_Layout.md -->
 
 <!-- toc -->
 
 - [Flash](#flash)
-- [Kernel code](#kernel-code)
-- [Process code](#process-code)
+  * [Kernel code](#kernel-code)
+  * [Process code](#process-code)
 - [RAM](#ram)
+  * [Kernel RAM](#kernel-ram)
+  * [Process RAM](#process-ram)
 - [Hardware Implementations](#hardware-implementations)
   * [SAM4L](#sam4l)
     + [Flash](#flash-1)
     + [RAM](#ram-1)
 
 <!-- tocstop -->
 
-Tock is intended to run on Cortex-M microcontrollers, which have
-non-volatile flash memory (for code) and RAM (for stack and data) in a
-single address space. While the Cortex-M architecture specifies a
-high-level layout of the address space, the exact layout of Tock can
-differ from board to board. Most boards simply define the beginning and
-end of flash and SRAM in their `layout.ld` file and then include the
-[generic Tock memory map](../boards/kernel_layout.ld).
+Tock is intended to run on microcontrollers like the Cortex-M, which have
+non-volatile flash memory (for code) and RAM (for stack and data) in a single
+address space. While the Cortex-M architecture specifies a high-level layout of
+the address space, the exact layout of Tock can differ from board to board.
+Most boards simply define the beginning and end of flash and SRAM in their
+`layout.ld` file and then include the [generic Tock memory
+map](../boards/kernel_layout.ld).
 
-## Flash
+The kernel has the highest privileges and thus also full control, meaning the
+ability to access all parts of code. Processes on the other hand have specific
```

Comment:
This is technically true of course, particularly from an MPU point of view, but I think we should at least note that the rust type system in practice restricts what the kernel can do. For example, a capsule cannot access a process's memory.

Reply:
What are the exact restrictions of a capsule?

Reply:
They cannot use `unsafe`. So they can only access memory and objects that are explicitly passed to them, and the only process memory a capsule ever gets passed is through the `allow` system call. They can't, for example, create and use a pointer to arbitrary memory, even though they are executing as privileged code.

Reply:
Right, this is a bit tricky do decide whether it goes in this doc or not. It *is* important for the tock isolation model that capsules are restricted from accessing arbitrary process memroy--they can only access process memory the process explicitly shares with them (through the allow system call).

So maybe the point in this doc, from a process perspective is that a process is guaranteed that capsules or other processes cannot access any of their memory unless the process `allow`s it. That's a guarantee the process can rely on, rather than a restrictions.